File: desktop/sidecar/daemora_sidecar/daemora_llm.py
"""Custom LLM plugin — pipes LiveKit Agents' LLM interface into Daemora's
/api/chat SSE. Daemora becomes the "model", tools and all.

Targets livekit-agents 1.5.x API:
  - Subclass llm.LLM, override chat() → returns LLMStream
  - Subclass llm.LLMStream, override _run() to emit ChatChunks
  - ChatChunk(id=..., delta=ChoiceDelta(content=..., role="assistant", tool_calls=[]))
"""

from __future__ import annotations

import json
import logging
import uuid
from typing import Optional

import httpx
from livekit.agents import APIConnectOptions, llm

from .voice_config import VoiceConfig

log = logging.getLogger("daemora.voice.llm")


class DaemoraLLM(llm.LLM):
    """LiveKit LLM backed by Daemora's /api/chat SSE stream.

    Each user turn becomes one Daemora task. Daemora runs its full agent
    loop (tools, memory, crew, MCP), and we surface the text:delta events
    as LLM tokens back to the voice pipeline.
    """

    def __init__(self, cfg: VoiceConfig, session_id: str = "main"):
        super().__init__()
        self._cfg = cfg
        # Voice shares the "main" session with text chat by default, so the
        # transcript + Daemora's reply stream live into the chat UI exactly
        # like typing does. Override only if you want an isolated voice
        # history (e.g. multi-user voice rooms later).
        self._session_id = session_id
        self._client = httpx.AsyncClient(timeout=120.0)

    @property
    def provider(self) -> str:
        return "daemora"

    @property
    def model(self) -> str:
        return "daemora-agent"

    async def aclose(self):
        await self._client.aclose()

    def chat(
        self,
        *,
        chat_ctx: llm.ChatContext,
        tools: list | None = None,
        conn_options: APIConnectOptions = APIConnectOptions(),
        parallel_tool_calls=None,
        tool_choice=None,
        extra_kwargs=None,
    ) -> "DaemoraLLMStream":
        # Pick the most recent user message — that's what we POST to Daemora.
        user_input = ""
        for item in reversed(list(chat_ctx.items)):
            if getattr(item, "role", None) == "user":
                content = item.content
                if isinstance(content, list):
                    user_input = " ".join(str(c) for c in content if c)
                else:
                    user_input = str(content or "")
                if user_input:
                    break
        return DaemoraLLMStream(
            llm=self,
            chat_ctx=chat_ctx,
            tools=tools or [],
            conn_options=conn_options,
            user_input=user_input,
        )


class DaemoraLLMStream(llm.LLMStream):
    def __init__(
        self,
        *,
        llm: DaemoraLLM,
        chat_ctx: llm.ChatContext,
        tools: list,
        conn_options: APIConnectOptions,
        user_input: str,
    ):
        super().__init__(llm=llm, chat_ctx=chat_ctx, tools=tools, conn_options=conn_options)
        self._user_input = user_input

    async def _run(self) -> None:
        cfg: VoiceConfig = self._llm._cfg  # type: ignore[attr-defined]
        client = self._llm._client  # type: ignore[attr-defined]
        request_id = str(uuid.uuid4())

        if not self._user_input.strip():
            log.debug("empty user input, nothing to stream")
            return

        headers = {"Content-Type": "application/json"}
        if cfg.daemora_auth_token:
            headers["Authorization"] = f"Bearer {cfg.daemora_auth_token}"

        # Send voice: true flag — Daemora appends voice instructions to the
        # system prompt at runtime, no token waste on user input.
        try:
            r = await client.post(
                f"{cfg.daemora_http}/api/chat",
                headers=headers,
                json={
                    "input": self._user_input,
                    "sessionId": self._llm._session_id,  # type: ignore[attr-defined]
                    "voice": True,
                },
            )
            r.raise_for_status()
            task_id = r.json().get("taskId")
        except Exception as e:
            log.error("Daemora /api/chat POST failed: %s", e)
            return

        if not task_id:
            log.error("Daemora returned no taskId")
            return

        # 2) Tail the task SSE stream; emit text:delta as ChatChunk
        sse_url = f"{cfg.daemora_http}/api/tasks/{task_id}/stream"
        sse_headers = dict(headers)
        sse_headers["Accept"] = "text/event-stream"

        try:
            async with client.stream("GET", sse_url, headers=sse_headers, timeout=None) as resp:
                event_name: Optional[str] = None
                async for line in resp.aiter_lines():
                    line = line.strip()
                    if not line:
                        event_name = None
                        continue
                    if line.startswith("event:"):
                        event_name = line.split(":", 1)[1].strip()
                        continue
                    if not line.startswith("data:"):
                        continue
                    data_str = line.split(":", 1)[1].strip()
                    try:
                        payload = json.loads(data_str)
                    except Exception:
                        continue

                    if event_name == "text:delta":
                        delta = payload.get("delta", "")
                        if not delta:
                            continue
                        self._event_ch.send_nowait(
                            llm.ChatChunk(
                                id=request_id,
                                delta=llm.ChoiceDelta(
                                    role="assistant",
                                    content=delta,
                                    tool_calls=[],
                                ),
                            )
                        )
                    elif event_name in ("task:completed", "task:failed", "text:end"):
                        return
        except Exception as e:
            log.error("Daemora SSE stream error: %s", e)
            return

File: src/voice/DaemoraLLM.ts
/**
 * DaemoraLLM — LiveKit `llm.LLM` adapter that streams agent responses
 * from the main Daemora HTTP process over SSE.
 *
 * The LiveKit voice pipeline runs in a child worker process (rtc-node
 * crashes under tsx, so it has to be compiled + spawned separately).
 * That means the LLM bridge cannot hold a direct AgentLoop reference —
 * it has to talk to the running main process over HTTP.
 *
 * Protocol:
 *   1. POST /api/chat { input, sessionId: "main" } → { taskId }
 *   2. Subscribe to GET /api/tasks/:taskId/stream (SSE) and forward
 *      every `text:delta` event as a ChatChunk delta.
 *   3. Close the LLMStream queue on `task:state = completed|failed`.
 *
 * Because the web chat and voice share the same session id ("main"),
 * the conversation history is unified — a voice turn appears in the
 * web chat UI and vice versa.
 */

import { randomUUID } from "node:crypto";

import type { APIConnectOptions } from "@livekit/agents";
import { llm } from "@livekit/agents";

// Session shared with web chat + every channel so history is unified.
const VOICE_SESSION_ID = "main";

export interface DaemoraLLMOpts {
  /** Base URL of the main Daemora HTTP server. */
  readonly daemoraUrl: string;
  /** Optional model id (purely for `model` getter — routing happens on the server). */
  readonly model?: string;
}

export class DaemoraLLM extends llm.LLM {
  readonly daemoraUrl: string;
  readonly modelId: string;

  constructor(opts: DaemoraLLMOpts) {
    super();
    this.daemoraUrl = opts.daemoraUrl.replace(/\/$/, "");
    this.modelId = opts.model ?? "daemora";
  }

  override label(): string {
    return "daemora.LLM";
  }

  override get model(): string {
    return this.modelId;
  }

  override get provider(): string {
    return "daemora";
  }

  override chat({
    chatCtx,
    toolCtx,
    connOptions,
  }: {
    chatCtx: llm.ChatContext;
    toolCtx?: llm.ToolContext;
    connOptions?: APIConnectOptions;
    parallelToolCalls?: boolean;
    toolChoice?: llm.ToolChoice;
    extraKwargs?: Record<string, unknown>;
  }): llm.LLMStream {
    return new DaemoraLLMStream(this, {
      chatCtx,
      ...(toolCtx ? { toolCtx } : {}),
      ...(connOptions ? { connOptions } : {}),
    });
  }
}

class DaemoraLLMStream extends llm.LLMStream {
  private readonly daemora: DaemoraLLM;

  constructor(
    daemora: DaemoraLLM,
    opts: {
      chatCtx: llm.ChatContext;
      toolCtx?: llm.ToolContext;
      connOptions?: APIConnectOptions;
    },
  ) {
    const connOptions = opts.connOptions ?? ({ maxRetry: 0, timeoutMs: 120_000 } as APIConnectOptions);
    if (opts.toolCtx) {
      super(daemora, { chatCtx: opts.chatCtx, toolCtx: opts.toolCtx, connOptions });
    } else {
      super(daemora, { chatCtx: opts.chatCtx, connOptions });
    }
    this.daemora = daemora;
  }

  protected override async run(): Promise<void> {
    const userText = extractLatestUserText(this.chatCtx);
    const chunkId = randomUUID();

    if (!userText) {
      this.queue.close();
      return;
    }

    try {
      // Kick off the task.
      const kickoff = await fetch(`${this.daemora.daemoraUrl}/api/chat`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          input: userText,
          sessionId: VOICE_SESSION_ID,
          // Flag: ask the server to add a "voice mode" prelude to the
          // system prompt so the reply is spoken-style (no markdown,
          // 1–3 short sentences). TTS reads bullet points and pipe
          // characters out loud otherwise.
          voiceMode: true,
        }),
        signal: this.abortController.signal,
      });
      if (!kickoff.ok) {
        throw new Error(`/api/chat ${kickoff.status}`);
      }
      const { taskId } = (await kickoff.json()) as { taskId: string };

      // Stream events for that task.
      await this.consumeSSE(`${this.daemora.daemoraUrl}/api/tasks/${taskId}/stream`, chunkId);
    } catch (err) {
      if (!this.abortController.signal.aborted) {
        // Surface the error up to the pipeline so it can recover.
        throw err;
      }
    } finally {
      this.queue.close();
    }
  }

  /**
   * Minimal SSE parser — reads `event:` / `data:` lines and routes
   * text deltas to the LLMStream queue. Terminates on `task:state`
   * with a completed / failed status, or when the stream ends.
   */
  private async consumeSSE(url: string, chunkId: string): Promise<void> {
    const resp = await fetch(url, {
      method: "GET",
      headers: { Accept: "text/event-stream" },
      signal: this.abortController.signal,
    });
    if (!resp.ok || !resp.body) throw new Error(`SSE ${resp.status}`);

    const reader = resp.body.getReader();
    const decoder = new TextDecoder();
    let buf = "";
    let currentEvent = "";

    while (true) {
      if (this.abortController.signal.aborted) {
        try { await reader.cancel(); } catch {}
        return;
      }
      const { done, value } = await reader.read();
      if (done) return;
      buf += decoder.decode(value, { stream: true });

      let idx;
      while ((idx = buf.indexOf("\n")) >= 0) {
        const line = buf.slice(0, idx).replace(/\r$/, "");
        buf = buf.slice(idx + 1);

        if (line.startsWith("event:")) {
          currentEvent = line.slice(6).trim();
        } else if (line.startsWith("data:")) {
          const dataRaw = line.slice(5).trim();
          if (!dataRaw) continue;
          let payload: unknown;
          try { payload = JSON.parse(dataRaw); } catch { continue; }

          if (currentEvent === "text:delta") {
            const delta = (payload as { delta?: string }).delta;
            if (delta) {
              this.queue.put({
                id: chunkId,
                delta: { role: "assistant", content: delta },
              });
            }
          } else if (currentEvent === "task:state") {
            const status = (payload as { status?: string }).status;
            if (status === "completed" || status === "failed") return;
          }
        }
        // blank line → dispatch boundary; nothing to do here.
      }
    }
  }
}

function extractLatestUserText(ctx: llm.ChatContext): string {
  const items = ctx.items;
  for (let i = items.length - 1; i >= 0; i--) {
    const item = items[i];
    if (!item) continue;
    if (item.type !== "message") continue;
    const msg = item as llm.ChatMessage;
    if (msg.role !== "user") continue;
    const text = msg.textContent;
    if (text && text.trim()) return text.trim();
  }
  return "";
}
